File: backend/reports.py
"""AD_CONVERSION StatReport 기반 구매완료 전환 집계.

/stats의 convAmt는 장바구니 매출이 포함돼서, 구매완료만 분리하려면
일자별 AD_CONVERSION 리포트를 받아 goalTp=purchase 행만 합산해야 함.
"""
import base64
import hashlib
import hmac
import time
from collections import defaultdict
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import date, timedelta
from urllib.parse import urlparse

import requests

from .naver_api import NaverAdsClient

REPORT_TP = "AD_CONVERSION"
POLL_INTERVAL = 2.0
POLL_MAX = 30
TTL_PAST = 60 * 60 * 24   # 확정일: 24시간
TTL_TODAY = 60 * 5        # 오늘: 5분

# (customer_id, date_iso) -> (fetched_ts, rows)
# rows: list of dicts with keys: campaignId, adgroupId, keywordId, adId, goalTp, cnt, amt
_date_cache: dict[tuple[str, str], tuple[float, list[dict]]] = {}


def _ttl_for(d: date) -> int:
    return TTL_TODAY if d >= date.today() else TTL_PAST


def _cache_get(customer_id: str, d: date) -> list[dict] | None:
    hit = _date_cache.get((customer_id, d.isoformat()))
    if not hit:
        return None
    ts, rows = hit
    if time.time() - ts > _ttl_for(d):
        return None
    return rows


def _cache_set(customer_id: str, d: date, rows: list[dict]) -> None:
    _date_cache[(customer_id, d.isoformat())] = (time.time(), rows)


def _download_report(client: NaverAdsClient, download_url: str) -> str:
    parsed = urlparse(download_url)
    uri = parsed.path  # path만, 쿼리 제외
    ts = str(int(time.time() * 1000))
    msg = f"{ts}.GET.{uri}"
    sig = base64.b64encode(
        hmac.new(client.secret_key.encode(), msg.encode(), hashlib.sha256).digest()
    ).decode()
    headers = {
        "X-Timestamp": ts,
        "X-API-KEY": client.api_key,
        "X-Customer": client.customer_id,
        "X-Signature": sig,
    }
    r = requests.get(download_url, headers=headers, timeout=30)
    r.raise_for_status()
    return r.text


def _fetch_one_date(client: NaverAdsClient, d: date) -> list[dict]:
    stat_dt = f"{d.isoformat()}T00:00:00.000Z"
    job = client.post("/stat-reports", {"reportTp": REPORT_TP, "statDt": stat_dt})
    job_id = job.get("reportJobId") or job.get("id")
    status = job.get("status")

    # 폴링
    info = job
    for _ in range(POLL_MAX):
        if status in ("BUILT", "DONE", "NONE"):
            break
        if status in ("FAILED", "REGIST_ERROR"):
            raise RuntimeError(f"report failed for {d}: {info}")
        time.sleep(POLL_INTERVAL)
        info = client.get(f"/stat-reports/{job_id}")
        status = info.get("status")

    if status == "NONE":
        return []
    if status not in ("BUILT", "DONE"):
        raise RuntimeError(f"report timeout for {d}: status={status}")

    download_url = info.get("downloadUrl")
    if not download_url:
        return []
    text = _download_report(client, download_url)

    rows = []
    for ln in text.strip().split("\n"):
        if not ln:
            continue
        cols = ln.split("\t")
        if len(cols) < 13:
            continue
        try:
            amt = float(cols[-1]) if cols[-1] else 0.0
            cnt = float(cols[-2]) if cols[-2] else 0.0
        except ValueError:
            continue
        rows.append({
            "campaignId": cols[2],
            "adgroupId": cols[3],
            "keywordId": cols[4],  # "-" 가능
            "adId": cols[5],
            "goalTp": cols[-3],
            "cnt": cnt,
            "amt": amt,
        })
    return rows


def get_purchase_rows(client: NaverAdsClient, since: date, until: date) -> list[dict]:
    """since ~ until (inclusive) 기간의 구매완료 전환 행만 반환.

    customer_id로 캐시 분리 + 일자별 병렬 fetch.
    """
    cid = client.customer_id
    dates: list[date] = []
    d = since
    while d <= until:
        dates.append(d)
        d += timedelta(days=1)

    misses = [d for d in dates if _cache_get(cid, d) is None]
    if misses:
        with ThreadPoolExecutor(max_workers=min(8, len(misses))) as ex:
            futs = {ex.submit(_fetch_one_date, client, d): d for d in misses}
            for fut in as_completed(futs):
                d = futs[fut]
                try:
                    rows = fut.result()
                except Exception as e:
                    print(f"  [report fetch fail] {d}: {e}", flush=True)
                    rows = []
                _cache_set(cid, d, rows)

    result: list[dict] = []
    for d in dates:
        rows = _cache_get(cid, d) or []
        for r in rows:
            if r["goalTp"] == "purchase":
                result.append(r)
    return result


def aggregate(rows: list[dict], level: str) -> dict[str, dict]:
    """level: 'campaign' | 'adgroup' | 'keyword' — id -> {cnt, amt}.

    keyword 레벨에서는 keywordId가 '-'(확장검색/소재 기반)인 행은 제외.
    """
    key_field = {"campaign": "campaignId", "adgroup": "adgroupId", "keyword": "keywordId"}[level]
    out: dict[str, dict] = defaultdict(lambda: {"cnt": 0.0, "amt": 0.0})
    for r in rows:
        k = r[key_field]
        if level == "keyword" and (not k or k == "-"):
            continue
        out[k]["cnt"] += r["cnt"]
        out[k]["amt"] += r["amt"]
    return dict(out)

File: web/src/lib/reports.ts
// AD_CONVERSION StatReport 기반 구매완료 전환 집계. backend/reports.py 포팅.
// /stats의 convAmt는 장바구니 포함이라, 구매완료만 분리하려면 일자별 리포트 필요.
import { createHmac } from "node:crypto";
import { naver, type NaverCreds } from "./naver";

const REPORT_TP = "AD_CONVERSION";
const POLL_INTERVAL_MS = 2000;
const POLL_MAX = 30;
const TTL_PAST_MS = 60 * 60 * 24 * 1000;
const TTL_TODAY_MS = 60 * 5 * 1000;

export type PurchaseRow = {
  campaignId: string; adgroupId: string; keywordId: string; adId: string;
  goalTp: string; cnt: number; amt: number;
};

type CacheEntry = { ts: number; rows: PurchaseRow[] };
// key: `${customerId}:${YYYY-MM-DD}`. 프로세스 로컬 캐시 — Vercel 서버리스에선 동일 instance 유지 동안만 유효.
const dateCache = new Map<string, CacheEntry>();

function todayIso(): string {
  return new Date().toISOString().slice(0, 10);
}

function ttlFor(dateIso: string): number {
  return dateIso >= todayIso() ? TTL_TODAY_MS : TTL_PAST_MS;
}

function cacheGet(customerId: string, dateIso: string): PurchaseRow[] | null {
  const hit = dateCache.get(`${customerId}:${dateIso}`);
  if (!hit) return null;
  if (Date.now() - hit.ts > ttlFor(dateIso)) return null;
  return hit.rows;
}

function cacheSet(customerId: string, dateIso: string, rows: PurchaseRow[]) {
  dateCache.set(`${customerId}:${dateIso}`, { ts: Date.now(), rows });
}

type ReportJob = { reportJobId?: string | number; id?: string | number; status?: string; downloadUrl?: string };

async function downloadReport(creds: NaverCreds, downloadUrl: string): Promise<string> {
  const u = new URL(downloadUrl);
  const uri = u.pathname;
  const ts = String(Date.now());
  const msg = `${ts}.GET.${uri}`;
  const sig = createHmac("sha256", creds.secretKey).update(msg).digest("base64");
  const r = await fetch(downloadUrl, {
    headers: {
      "X-Timestamp": ts,
      "X-API-KEY": creds.apiKey,
      "X-Customer": creds.customerId,
      "X-Signature": sig,
    },
    cache: "no-store",
  });
  if (!r.ok) throw new Error(`report download ${r.status}`);
  return r.text();
}

function parseReport(text: string): PurchaseRow[] {
  const rows: PurchaseRow[] = [];
  for (const ln of text.trim().split("\n")) {
    if (!ln) continue;
    const cols = ln.split("\t");
    if (cols.length < 13) continue;
    const amt = parseFloat(cols[cols.length - 1]) || 0;
    const cnt = parseFloat(cols[cols.length - 2]) || 0;
    rows.push({
      campaignId: cols[2],
      adgroupId: cols[3],
      keywordId: cols[4],
      adId: cols[5],
      goalTp: cols[cols.length - 3],
      cnt,
      amt,
    });
  }
  return rows;
}

async function fetchOneDate(creds: NaverCreds, dateIso: string): Promise<PurchaseRow[]> {
  const statDt = `${dateIso}T00:00:00.000Z`;
  let info = await naver.post<ReportJob>(creds, "/stat-reports", { reportTp: REPORT_TP, statDt });
  const jobId = info.reportJobId ?? info.id;
  let status = info.status;

  for (let i = 0; i < POLL_MAX; i++) {
    if (status === "BUILT" || status === "DONE" || status === "NONE") break;
    if (status === "FAILED" || status === "REGIST_ERROR") {
      throw new Error(`report failed ${dateIso}: ${status}`);
    }
    await new Promise((res) => setTimeout(res, POLL_INTERVAL_MS));
    info = await naver.get<ReportJob>(creds, `/stat-reports/${jobId}`);
    status = info.status;
  }

  if (status === "NONE") return [];
  if (status !== "BUILT" && status !== "DONE") {
    throw new Error(`report timeout ${dateIso}: ${status}`);
  }
  if (!info.downloadUrl) return [];
  return parseReport(await downloadReport(creds, info.downloadUrl));
}

function eachDate(since: string, until: string): string[] {
  const out: string[] = [];
  const s = new Date(since + "T00:00:00Z");
  const u = new Date(until + "T00:00:00Z");
  for (let d = s; d <= u; d = new Date(d.getTime() + 86400000)) {
    out.push(d.toISOString().slice(0, 10));
  }
  return out;
}

export async function getPurchaseRows(
  creds: NaverCreds, since: string, until: string,
): Promise<PurchaseRow[]> {
  const dates = eachDate(since, until);
  const misses = dates.filter((d) => cacheGet(creds.customerId, d) === null);

  if (misses.length) {
    // 일자별 병렬. Vercel 타임아웃 감안해 동시성 제한 없이 Promise.all.
    const results = await Promise.allSettled(misses.map((d) => fetchOneDate(creds, d)));
    results.forEach((res, i) => {
      const d = misses[i];
      cacheSet(creds.customerId, d, res.status === "fulfilled" ? res.value : []);
    });
  }

  const out: PurchaseRow[] = [];
  for (const d of dates) {
    for (const r of cacheGet(creds.customerId, d) ?? []) {
      if (r.goalTp === "purchase") out.push(r);
    }
  }
  return out;
}

export type AggLevel = "campaign" | "adgroup" | "keyword";
export function aggregate(rows: PurchaseRow[], level: AggLevel): Record<string, { cnt: number; amt: number }> {
  const keyField: keyof PurchaseRow = level === "campaign" ? "campaignId" : level === "adgroup" ? "adgroupId" : "keywordId";
  const out: Record<string, { cnt: number; amt: number }> = {};
  for (const r of rows) {
    const k = r[keyField] as string;
    if (level === "keyword" && (!k || k === "-")) continue;
    if (!out[k]) out[k] = { cnt: 0, amt: 0 };
    out[k].cnt += r.cnt;
    out[k].amt += r.amt;
  }
  return out;
}
